src/store.c

```diff
@@ -339,112 +375,190 @@ static void writer_close(
     close(store->fd);
 }
 
+static void init_store_offsets(
+    struct rill_store* store, size_t vals, size_t inverse_vals)
+{
+    store->head->index_a_off = sizeof(struct header);
+    store->head->index_b_off = store->head->index_a_off + indexer_cap(inverse_vals);
+    store->head->data_a_off = store->head->index_b_off + indexer_cap(vals);
+
+    store->index_a = (void *) ((uintptr_t) store->vma + store->head->index_a_off);
+    store->index_b = (void *) ((uintptr_t) store->vma + store->head->index_b_off);
+    store->data_a = (void *) ((uintptr_t) store->vma + store->head->data_a_off);
+}
+
+static void prepare_col_b_offsets(
+    struct rill_store* store, struct encoder* coder_a)
+{
+    store->head->data_b_off = store->head->data_a_off + coder_off(coder_a);
+    store->data_b = (void *) ((uintptr_t) store->vma + store->head->data_b_off);
+}
+
 bool rill_store_write(
         const char *file,
-        rill_ts_t ts, size_t quant,
+        rill_ts_t ts,
+        size_t quant,
         struct rill_pairs *pairs)
 {
     rill_pairs_compact(pairs);
     if (!pairs->len) return true;
 
-    struct vals *vals = vals_from_pairs(pairs);
+    struct vals *vals = vals_cols_from_pairs(pairs, rill_col_b);
     if (!vals) goto fail_vals;
+    struct vals *invert_vals = vals_cols_from_pairs(pairs, rill_col_a);
+    if (!invert_vals) goto fail_invert_vals;
 
     struct rill_store store = {0};
-    if (!writer_open(&store, file, vals, pairs->len, ts, quant)) {
+    if (!writer_open(&store, file, vals, invert_vals,
+                     pairs->len, ts, quant)) {
         rill_fail("unable to create '%s'", file);
         goto fail_open;
     }
 
-    struct indexer *indexer = indexer_alloc(pairs->len);
-    struct encoder coder = writer_begin(&store, vals, indexer);
+    struct indexer *indexer_a = indexer_alloc(invert_vals->len);
+    if (!indexer_a) goto fail_indexer_a_alloc;
+    struct indexer *indexer_b = indexer_alloc(vals->len);
+    if (!indexer_b) goto fail_indexer_b_alloc;
```

Comment:
I'm pretty sure I've mentioned this before so apologies if I'm getting annoying. `indexer_alloc` should not really be required anymore as its in-memory representation is the same as its in region representation. On top of that I've looked in both `store_write` and `store_merge` and in both cases the region is properly setup by the time you need the indexes. So, unless I'm missing something, it looks like you should be able to use the memory region directly.